File: lib/generate_guangna_sets.py
#!/usr/bin/env python3
"""
generate_guangna_sets.py
─────────────────────────────────────────────
Regenerates the GUANGNA_SETS dict literal for pbn_guangna.py from the
canonical guangna-sets.json -- companion to generate_guangna_colors.py,
same reasoning: this is the single source of truth for which codes
belong to which marker set, shared between the Python backend
(pbn-webservice) and the TypeScript frontend (crea-bea-studio).

Run this, then paste the printed output over the existing GUANGNA_SETS
block in pbn_guangna.py, whenever guangna-sets.json changes.

Usage: python3 generate_guangna_sets.py > guangna_sets_output.txt
"""

import json

def main():
    with open("guangna-sets.json") as f:
        data = json.load(f)

    print("GUANGNA_SETS = {")
    for label, codes in data.items():
        codes_str = ", ".join(f'"{c}"' for c in codes)
        print(f'    "{label}": [{codes_str}],')
    print("}")
    print()
    print("GUANGNA_SET_OPTIONS = list(GUANGNA_SETS.keys())")

if __name__ == "__main__":
    main()

File: lib/generate-guangna-sets.mjs
// generate-guangna-sets.mjs
// ─────────────────────────────────────────────
// Regenerates the GUANGNA_SETS object literal for lib/guangna.ts from
// the canonical guangna-sets.json -- companion to
// generate-guangna-colors.mjs, same reasoning.
//
// Run this, then paste the printed output over the existing
// GUANGNA_SETS block in lib/guangna.ts, whenever guangna-sets.json
// changes. Note: the "GN.8101-366 (366 colors)" entry is written out
// as an explicit array here (not "Object.keys(GN_COLORS)" like the
// original hand-written file did) -- functionally identical, since the
// canonical JSON's 366 set is already exactly GN_COLORS' key order,
// but this avoids the generator needing to know about GN_COLORS at all.
//
// Usage: node generate-guangna-sets.mjs > guangna-sets-output.txt

import { readFileSync } from "fs";

const data = JSON.parse(readFileSync("./guangna-sets.json", "utf8"));

let out = "export const GUANGNA_SETS: Record<string,string[]> = {\n";
for (const [label, codes] of Object.entries(data)) {
  const codesStr = codes.map((c) => `"${c}"`).join(",");
  out += `  "${label}":[${codesStr}],\n`;
}
out += "};\n";

console.log(out);
